LanguageFeatures/nnbd/const_evaluation_A10_t02.dart

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Two objects T1 and T2 which are instances of Type (that is,
+/// runtime type objects) are considered equal if and only if the runtime type
+/// objects T1 and T2 corresponds to the types S1 and S2 respectively, and the
+/// normal forms NORM(S1) and NORM(S2) are syntactically equal up to equivalence
+/// of bound variables and ignoring * modifiers on types.
+/// ...
+/// Two constant type literals T1 and T2 compare as identical if they are equal
+/// using the definition of runtime type equality specified above.
+///
+/// @description Checks that two constant type literals `T1` and `T2` are not
+/// equals if they are not equal according to the runtime type equality.
+///
+/// @author sgrekhov22@gmail.com
+
+import "dart:async";
+import "../../Utils/expect.dart";
+
+typedef NullableObject = Object?;
+typedef Void = void;
+
+main() {
+  Expect.notEquals(Null, dynamic);
+  Expect.notEquals(Null, Never);
+  Expect.notEquals(Null, Object);
+  Expect.notEquals(Null, NullableObject);
+  Expect.notEquals(Null, Void);
+  Expect.notEquals(Null, FutureOr<Null>);
+
+  Expect.notEquals(dynamic, Never);
+  Expect.notEquals(dynamic, Object);
+  Expect.notEquals(dynamic, NullableObject);
+  Expect.notEquals(dynamic, Void);
+
+  Expect.notEquals(Never, Object);
+  Expect.notEquals(Never, NullableObject);
+  Expect.notEquals(Never, Void);
+  Expect.notEquals(Never, FutureOr<Never>);
+
+  Expect.notEquals(Object, NullableObject);
+  Expect.notEquals(Object, Void);
+
+  Expect.notEquals(Void, NullableObject);
+
+  Expect.notEquals(NullableObject, FutureOr<Null>);
```

Comment:
We're testing the `==` operator on reified types, and this means that it does matter which operand comes first. So we should have symmetric coverage: If we're testing `A == B` then we should also test `B == A` (in both cases, expecting `false`, but that's handled by `notEquals`, we just need to know that it's all about `==`, and hence symmetry doesn't come for free, and it can have bugs).

This basically means that we won't have shorter and shorter blocks (`Null` with everyone, followed by `dynamic` with everyone except `Null`, etc. becomes `Null` with everyone, `dynamic` with everyone, etc.).

We can't possibly test all combinations of reified types that should be unequal, but we should test the pairs that are "likely to get it wrong".

We could certainly have all combinations of primitive top types plus `Object?`:

- (Object?, dynamic) // Missing
- (Object?, void) // Missing
- (dynamic, Object?) // Line 35
- (dynamic, void) // Line 36
- (void, Object?) // Line 46
- (void, dynamic) // Missing.

It is already tested that `Object` is unequal to the simple top types. We should test that `Null` is unequal to bottom types (`Never`/`X extends Never`).

It doesn't hurt to test that `Never` and `void` are different, but for those (wildly different) types we can rely on existing practice (which would have been broken a gazillion times if those weren't recognized as different). So it's perfectly fine that we just have handful of those wildly different pairs.

Finally, we should have some composite types with those those "near-equal pairs" that I just tried to outline: `List<void> != List<dynamic>`, `void Function(dynamic) != void Function(void)` (the toughest test for inequality is the case where just one component differs, whereas the toughest test for `identical` is the case where everything changes, but only in ways that are eliminated by normalization).

Reply:
OK, when this one has been adjusted to cover some extra cases, LanguageFeatures/nnbd/const_evaluation_A10_t03.dart can be adjusted to test the same type pairs wrapped in `Set<_>`. And there could be yet another test containing the cases about function types.

Reply:
Added